stdlib/universe/narrow_state_tracking.go

```diff
@@ -0,0 +1,171 @@
+	a := &narrowStateTrackingTransformationAdapter{t: t}
+	nt, d, err := execute.NewNarrowStateTransformation(id, a, mem)
+	if err != nil {
+		return nil, nil, err
+	}
+	return nt, d, nil
+}
+
+type narrowStateTrackingTransformationAdapter struct {
+	t *stateTrackingTransformation
+}
+
+type trackedState struct {
+	start,
+	prevTime values.Time
+
+	countInState,
+	durationInState bool
+
+	count,
+	duration int64
+}
+
+func (a *narrowStateTrackingTransformationAdapter) Process(chunk table.Chunk, state interface{}, d *execute.TransportDataset, mem memory.Allocator) (interface{}, bool, error) {
+	// Track whether or not the state has been modified
+	mod := false
+
+	if chunk.Len() == 0 {
+		return state, mod, nil
+	}
+
+	if state == nil {
+		state = trackedState{
+			count:    -1,
+			duration: -1,
+		}
+		mod = true
+	}
+	s := state.(trackedState)
+
+	cols := chunk.Cols()
+	fn, err := a.t.fn.Prepare(cols)
+	if err != nil {
+		return s, mod, err
+	}
+
+	// Get the index for the time column
+	timeIdx := execute.ColIdx(a.t.timeCol, cols)
+	if timeIdx < 0 {
+		// Do we need to retain and process the unmodified chunk
+		// before an early return?
+		//
+		// chunk.Retain()
+		// d.Process(chunk)
+		return s, mod, errors.Newf(codes.FailedPrecondition, "column %q does not exist", a.t.timeCol)
+	}
+
+	stateCounts := array.NewIntBuilder(mem)
+	stateDurations := array.NewIntBuilder(mem)
+
+	nrows := chunk.Len()
+
+	stateCounts.Resize(nrows)
+	stateDurations.Resize(nrows)
+
+	for i := 0; i < nrows; i++ {
+		// Evaluate the predicate for the current row
+		match, err := fn.EvalRow(a.t.ctx, i, chunk)
+		if err != nil {
+			return s, mod, err
+		}
+
+		// Update state duration
+		if a.t.durationColumn != "" {
+			times := chunk.Times(timeIdx)
+			if times.IsNull(i) {
+				return s, mod, errors.New(codes.FailedPrecondition, "got a null timestamp")
+			}
+			ts := values.Time(times.Value(i))
+			if s.prevTime > ts {
+				return s, mod, errors.New(codes.FailedPrecondition, "got an out-of-order timestamp")
+			}
+			s.prevTime = ts
+			mod = true
+
+			if match {
+				if !s.durationInState {
+					s.durationInState = true
+					s.start = ts
+					s.duration = 0
+				} else {
+					s.duration = int64(ts - s.start)
+					if a.t.durationUnit > 0 {
+						s.duration = s.duration / a.t.durationUnit
+					}
+				}
+			} else {
+				s.durationInState = false
+				s.duration = -1
+			}
+			stateDurations.Append(s.duration)
+		}
+
+		// Update state count
+		if a.t.countColumn != "" {
+			if match {
+				if !s.countInState {
+					s.countInState = true
+					s.count = 1
+				} else {
+					s.count++
+				}
+			} else {
+				s.countInState = false
+				s.count = -1
+			}
+			mod = true
+			stateCounts.Append(s.count)
+		}
+	}
+
+	ncols := chunk.NCols()
+	newCols := make([]flux.ColMeta, 0, ncols+2)
+	newCols = append(newCols, cols...)
+
+	vs := make([]array.Interface, 0, ncols+2)
+	for i := 0; i < ncols; i++ {
+		vs = append(vs, chunk.Values(i))
```

Comment:
You will need to retain these arrays before copying them over.

Reply:
This is the proper place to call `Retain()` rather than on the chunk itself. This is mostly just because it's easier to validate by eye that the retain is proper. You can see here, "Oh, the array was accessed and then it was stored so this retain was necessary."